cpp/src/arrow/compute/kernels/aggregate_basic.cc

```diff
@@ -754,6 +839,30 @@ void RegisterScalarAggregateBasic(FunctionRegistry* registry) {
                aggregate::CountInit, func.get());
   DCHECK_OK(registry->AddFunction(std::move(func)));
 
+  func = std::make_shared<ScalarAggregateFunction>(
+      "count_distinct", Arity::Unary(), &count_distinct_doc, &default_count_options);
+
+  // Takes any input, outputs int64 scalar
+  aggregate::AddCountDistinctKernel<Int8Type>(int8(), func.get());
+  aggregate::AddCountDistinctKernel<Int16Type>(int16(), func.get());
+  aggregate::AddCountDistinctKernel<Int32Type>(int32(), func.get());
+  aggregate::AddCountDistinctKernel<Date32Type>(date32(), func.get());
+  aggregate::AddCountDistinctKernel<Int64Type>(int64(), func.get());
```

Comment:
You are providing explicit input types because you need them for the [`ArrayDataInlineVisitor`](https://github.com/apache/arrow/blob/master/cpp/src/arrow/visitor_inline.h#L141) invoked via [`VisitorArrayDataInline`](https://github.com/apache/arrow/blob/master/cpp/src/arrow/visitor_inline.h#L326). Note that `ArrayDataInlineVisitor` only checks for the base type, so we can possibly register less number of kernels by leveraging [implicit casting](https://arrow.apache.org/docs/cpp/compute.html#implicit-casts).

Reply:
Or use `Type::PhysicalType` when instantiating the kernel class.

Reply:
Let me try to see if I can reduce the number of calls. I need the type for the Memo encoder.
I tried to use some of the generators in codegen_internal but it seems we only have generators for the Exec call of kernels, not for the initialization call.

Reply:
You could iterate through IntTypes or PrimitiveTypes or something, and just have a switch-case (essentially a hardcoded generator) on the type ID. Or use VisitTypeInline with some enable_if trickery.

Reply:
Ok now I understand better, but still I was thinking about these lines:

The amount of code to make this "prettier" (using visitor, enable_if, hard-coded generators, etc) is going to result in more LoC and abstractions. Are there any reasons why do we want to make this here? Perhaps I'm missing something.
Also, if we rely in implicit cast: Doesn't that make it worse? Because we will need to apply a casting every time the user wants to use a kernel. Explicit registration should result in faster kernel resolutions right?

Let me know what you think here please!
cc @edponce @pitrou @lidavidm

Reply:
I agree we probably don't need implicit casting. I think Antoine's suggestion is best. That would look like this (untested):

```cpp
template <typename Type>
void AddCountDistinctKernel(std::shared_ptr<DataType> type,
                            ScalarAggregateFunction* func) {
  using PhysicalType = typename Type::PhysicalType;
  AddAggKernel(KernelSignature::Make({type}, ValueDescr::Scalar(int64())),
               aggregate::CountDistinctInit<PhysicalType>, func);
}
```

Reply:
That said elsewhere Antoine points out SumLikeInit - something like that is probably ideal and I don't think it is appreciably harder to read (at least given that other kernels use similar infrastructure).

Reply:
Sure, but given that some of these kernels don't support lists, structs, i.e. more complex types; I think a refactor should be take place in the future and having explicit code is better/clear instead of use more (temporary) abstractions/solutions.

Reply:
btw @pitrou I resolved the other conversation because this thread have more ideas.

Reply:
Sure, that's reasonable. I think the key points here are just to not generate distinct kernels where it's not needed (e.g. Timestamp should use the same underlying impl as Int64, and frankly, Int64/UInt64 can/should be consolidated, right?) and to make sure things like zoned timestamps are properly supported. That said, consistency with the rest of the kernels is nice, even if the codebase is not perfect about being consistent throughout (and Eduardo has been identifying/filing cases like this) - at the very least, it may be worth putting up a follow-up JIRA. (Also because I think there's other things we may want to do, e.g. we could perhaps support dictionaries.)

Reply:
@lidavidm What types would I need to register if I want to use the `PhysicalType` snippet? I don't see how many lines of code I can cut off by using the `PhysicalType`. Thanks!

Reply:
We'd still register the same types, however, we'd be instructing the compiler to instantiate the kernel less (since e.g. TimestampType and Int64Type would both delegate to CountDistinctImpl<Int64Type>). You could also do something like `AddCountDistinctKernel<Int64Type>(uint64())` which would collapse the signed and unsigned implementations for each bit width.

Also now that I look at it, we should be able to support interval types here too - e.g. months interval is the same as int32, day-time interval is the same as int64.

Reply:
(I think this is why something like SumLikeInit is suggested, since you could work out some templates/helpers to collapse a lot of these cases, but yes, that would be more complicated.)

Reply:
I support trying to use `Type::PhysicalType` and if you need generator dispatcher...well instead of making a custom one you could add an `Exec` method that redirects to `ExecArray` or `ExecScalar` and then you could use the existing GDs.

Reply:
> We'd still register the same types, however, we'd be instructing the compiler to instantiate the kernel less (since e.g. TimestampType and Int64Type would both delegate to CountDistinctImpl)

I see thanks.

> make sure things like zoned timestamps are properly supported

How does the timezone gets encoded into the timestamp? is it part of its physical data layout? I think is a good idea to add a test for this case and for intervals as well ;)

> You could also do something like AddCountDistinctKernel<Int64Type>(uint64()) which would collapse the signed and unsigned implementations for each bit width.

Does this will use implicit casting for kernel resolution?

btw I can use an approach like SumLike no problem (is not that complicated), my point is that we will have more lines of code/abstractions at the end and we will make a potential refactor later to support more types.

> at the very least, it may be worth putting up a follow-up JIRA. (Also because I think there's other things we may want to do, e.g. we could perhaps support dictionaries.)

Yes that is part of my point too. I like to have consistency as well, but I think most of this refactor should take place once we identify we need to add other types and not before, is better to have a generic code but not to create internal API when we don't know yet the requirements for each case: i.e. when in doubt leave it out ;)

Reply:
> > We'd still register the same types, however, we'd be instructing the compiler to instantiate the kernel less (since e.g. TimestampType and Int64Type would both delegate to CountDistinctImpl)
> 
> I see thanks.
> 
> > make sure things like zoned timestamps are properly supported
> 
> How does the timezone gets encoded into the timestamp? is it part of its physical data layout? I think is a good idea to add a test for this case and for intervals as well ;)

It's a parameter of the type itself and is not in the data. So it's only relevant during things like type matching. Yes, adding tests would be a good idea.

> 
> > You could also do something like AddCountDistinctKernel(uint64()) which would collapse the signed and unsigned implementations for each bit width.
> 
> Does this will use implicit casting for kernel resolution?

No, so what we're talking about is registering the same kernel implementation for multiple types to save on generated code. Since for example, int64 and uint64 are basically the same from the perspective of this function and you can just treat one as the other.  

> 
> btw I can use an approach like SumLike no problem (is not that complicated), my point is that we will have more lines of code/abstractions at the end and we will make a potential refactor later to support more types.

I think the SumLike approach should handle potential future cases well. Yes, it does use more abstractions.

> 
> > at the very least, it may be worth putting up a follow-up JIRA. (Also because I think there's other things we may want to do, e.g. we could perhaps support dictionaries.)
> 
> Yes that is part of my point too. I like to have consistency as well, but I think most of this refactor should take place once we identify we need to add other types and not before, is better to have a generic code but not to create internal API when we don't know yet the requirements for each case: i.e. when in doubt leave it out ;)

Fair enough. Again, the approach here is likely OK, just needs some tweaking.